### PR TITLE
maven: Upgrade dependency versions

### PR DESCRIPTION
**Pull request by bhaisaab**
Updated most dependencies to latest minor releases, except:

- Gson 2.x
- Major spring framework version
- Servlet version
- Embedded jetty version
- Mockito version (beta)
- Mysql lib minor version upgrade (breaks mysql-ha plugin)

### CODE REVIEW

**Comment by bhaisaab** [comment]
Reopened from https://github.com/apache/cloudstack/pull/1393 with the branch on ASF remote. cc @DaanHoogland

**Comment by bhaisaab** [comment]
@DaanHoogland I've tested this and fix an issue, please feel free to make changes directly on this branch

**Comment by wido** [comment]
You also change Java code at the same time. Shouldn't this PR ONLY upgrade maven. Nothing more?

**Comment by DaanHoogland** [comment]
@wido not all updates are backwards compatible, I included the simpler changes for some.

**Comment by wido** [comment]
Ok, @DaanHoogland I understand.

Tests seem to fail though. Is that legit?

**Comment by DaanHoogland** [comment]
The travis tests, huh? haven't looked at those in a long time as these were failing always. Jenkins builds and the real test needed is integration tests.

**Comment by bhaisaab** [comment]
@bvbharat nice work, saw you ran BVT tests on #1415 (which was closed and merged here). Can you run and share results for this PR? Thanks.

**Comment by bvbharatk** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 97
 Hypervisor xenserver
 NetworkType Advanced
Passed=105
 Failed=9
 Skipped=4

**The follwing tests have known issues**
integration.smoke.test_iso.TestISO.test_07_list_default_iso 0
integration.smoke.test_privategw_acl.py 0
integration.smoke.test_vm_snapshots.TestSnapshots.test_01_test_vm_volume_snapshot 0
integration.smoke.test_vpc_vpn.TestVpcSite2SiteVpn.test_vpc_site2site_vpn 0
integration.smoke.test_iso.TestISO.test_04_extract_Iso 0


_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* integration.smoke.test_vm_snapshots.TestSnapshots

 * test_01_test_vm_volume_snapshot

* integration.smoke.test_privategw_acl.TestPrivateGwACL

 * test_02_vpc_privategw_static_routes

 * test_03_rvpc_privategw_static_routes

* integration.smoke.test_guest_vlan_range.TestDedicateGuestVlanRange

 * test_dedicateGuestVlanRange

* <nose.suite

 * ContextSuite context=TestNiciraContoller>:setup

* integration.smoke.test_primary_storage.TestPrimaryStorageServices

 * test_01_primary_storage_iscsi

* integration.smoke.test_templates.TestCreateTemplate

 * test_04_extract_template

* integration.smoke.test_iso.TestCreateIso

 * test_04_extract_Iso

 * test_07_list_default_iso


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_deploy_vgpu_enabled_vm
test_06_copy_template
test_06_copy_iso

**Passed test suits:**
integration.smoke.test_deploy_vm_with_userdata.TestDeployVmWithUserData
integration.smoke.test_affinity_groups_projects.TestDeployVmWithAffinityGroup
integration.smoke.test_portable_publicip.TestPortablePublicIPAcquire
integration.smoke.test_over_provisioning.TestUpdateOverProvision
integration.smoke.test_global_settings.TestUpdateConfigWithScope
integration.smoke.test_scale_vm.TestScaleVm
integration.smoke.test_service_offerings.TestCreateServiceOffering
integration.smoke.test_loadbalance.TestLoadBalance
integration.smoke.test_routers.TestRouterServices
integration.smoke.test_reset_vm_on_reboot.TestResetVmOnReboot
integration.smoke.test_snapshots.TestSnapshotRootDisk
integration.smoke.test_deploy_vms_with_varied_deploymentplanners.TestDeployVmWithVariedPlanners
integration.smoke.test_network.TestDeleteAccount
integration.smoke.test_non_contigiousvlan.TestUpdatePhysicalNetwork
integration.smoke.test_deploy_vm_iso.TestDeployVMFromISO
integration.smoke.test_public_ip_range.TestDedicatePublicIPRange
integration.smoke.test_multipleips_per_nic.TestDeployVM
integration.smoke.test_regions.TestRegions
integration.smoke.test_affinity_groups.TestDeployVmWithAffinityGroup
integration.smoke.test_network_acl.TestNetworkACL
integration.smoke.test_pvlan.TestPVLAN
integration.smoke.test_volumes.TestCreateVolume
integration.smoke.test_ssvm.TestSSVMs
integration.smoke.test_nic.TestNic
integration.smoke.test_deploy_vm_root_resize.TestDeployVM
integration.smoke.test_resource_detail.TestResourceDetail
integration.smoke.test_secondary_storage.TestSecStorageServices
integration.smoke.test_vm_life_cycle.TestDeployVM
integration.smoke.test_disk_offerings.TestCreateDiskOffering

**Comment by bhaisaab** [comment]
@bvbharat  Thanks for sharing, some failures could be due to env related issue. Please see if my comments below help you to improve your infra:

- Tests were run using Marvin 4.6.3-Snapshot. You should run tests against the Marvin built from the target branch after applying the PR patch, in this case it should be 4.9.0-Snapshot
- A test failure indicates VM snapshot not allowed on Xen, was the Xen server unclean?
- Test failure indicates VMs/Network were unreachable to public Internet, perhaps VMs are not allowed to ping external network/Internet; one of them indicated VLAN related issue
- A test failed as storagepool address lacked the scheme, nfs:// etc. (was null), check your marvin cfg
- A template and iso related test failed due to template extract operation failed, something wrong with template/url/file?
- The test_nicira_controller test failed due to issue in the test itself, a config key was None; perhaps you won't be able to test this without proper h/w and/or CloudStack configuration

**Comment by bhaisaab** [comment]
@swill rebased against latest master /cc @DaanHoogland

**Comment by DaanHoogland** [comment]
@bhaisaab I see you upgraded gson. Last time i tried that the integration tests started failing massively. I wasted a couple of day back then. did you verify this? gson is being used for internal communication of commands and has refactorred it's use of exclusion filters.

File: pom.xml [hunk 1]
```diff
@@ -65,58 +61,68 @@
...
-    <cs.jpa.version>2.1.0</cs.jpa.version>
+    <cs.bcprov.version>1.54</cs.bcprov.version>
+    <cs.jsch.version>0.1.53</cs.jsch.version>
+    <cs.jpa.version>2.1.1</cs.jpa.version>
     <cs.jasypt.version>1.9.2</cs.jasypt.version>
-    <cs.trilead.version>1.0.0-build217</cs.trilead.version>
-    <cs.ehcache.version>2.6.9</cs.ehcache.version>
+    <cs.trilead.version>1.0.0-build220</cs.trilead.version>
+    <cs.ehcache.version>2.6.11</cs.ehcache.version>
     <cs.gson.version>1.7.2</cs.gson.version>
```
Comment (bhaisaab) [inline pom.xml]: @DaanHoogland see here ^^, gson version is not changed at all

Comment (DaanHoogland) [inline pom.xml]: I am getting blind. Don't mind me for a while. I'm useless

Comment (bhaisaab) [inline pom.xml]: @DaanHoogland :) 
Perhaps my PR message confused you? Here I mean to say that we're not updating following:

> Updated most dependencies to latest minor releases, except:
> 
> Gson 2.x
> Major spring framework version
> Servlet version
> Embedded jetty version
> Mockito version (beta)
> Mysql lib minor version upgrade (breaks mysql-ha plugin)

Comment (DaanHoogland) [inline pom.xml]: yes, i should have caught the 'except' but I also could have sworn seeing a '2' near 'gson' I think I confused it with the saml version.

**Comment by swill** [comment]
There is a lot going on here and I don't like that Travis is failing since it tends to be pretty stable.  Also, Travis is failing with a "Pull Request #1397 maven: Upgrade dependency versions" error, which seems relevant since there are ton of packages changes in this update.  

I would like to better understand why Travis is failing this way.  @jburwell do you have thoughts on this?

**Comment by bhaisaab** [comment]
@swill Travis has been failing for several PRs when the test exceed the timeout period (30mins) https://travis-ci.org/apache/cloudstack/jobs/117688064
This has been going on for a while, we'll need to survey the set of smoke tests included in each run and optimize them (esp the global settings etc)

**Comment by jburwell** [comment]
@swill I agree there a lot of incongruent changes in this PR.  I would like to see it decomposed.  For example, the Bouncy Castle upgrade should be completely separated into a dedicated PR.  I would also like to understand what is motivating many of the dependency upgrades.

**Comment by bhaisaab** [comment]
@swill I've fixed for Travis now, locally I was successfully able to run more than 50 smoke/component tests against simulator. The Travis test result printer is fixed now and will also show the 'time taken' in seconds in the tabular output.

**Comment by bvbharatk** [comment]
@bhaisaab 

apologies for the late reply, I was busy working on some of the internal work and did not notice you comment. I am aware of some of these issues. Most of them are test data related,  I will fix them as soon as possible.
Thanks for your suggestions.

**Comment by bhaisaab** [comment]
@bvbharat np, looking forward to improvements

**Comment by runseb** [comment]
+1 to all the Travis changes based on code review and verification of travis tests. I would have put those changes in a separate PR however.

**Comment by swill** [comment]
This is failing to build in my CI environment.  The following tests are causing the build to fail.

It is failing here:
```
[INFO] Apache CloudStack Framework - Event Notification .. SUCCESS [0.485s]
[INFO] Apache CloudStack Cloud Engine Schema Component ... SUCCESS [4.719s]
[INFO] Apache CloudStack Framework - Jobs ................ SUCCESS [2.225s]
[INFO] Apache CloudStack Cloud Engine Internal Components API  SUCCESS [0.951s]
[INFO] Apache CloudStack Server .......................... FAILURE [47.577s]
[INFO] Apache CloudStack Framework - Quota ............... SKIPPED
[INFO] Apache CloudStack Usage Server .................... SKIPPED
[INFO] Apache CloudStack Cloud Engine Orchestration Component  SKIPPED
```

With these failed tests:
```
Tests run: 13, Failures: 2, Errors: 1, Skipped: 0, Time elapsed: 0.348 sec <<< FAILURE! - in org.apache.cloudstack.network.lb.CertServiceTest
runUploadSslCertWithCAChain(org.apache.cloudstack.network.lb.CertServiceTest)  Time elapsed: 0.026 sec  <<< ERROR!
com.cloud.utils.exception.CloudRuntimeException: Error parsing certificate data Parsing certificate/key failed: Could not parse certificate: java.io.IOException: Empty input
        at org.apache.cloudstack.network.lb.CertServiceImpl.uploadSslCert(CertServiceImpl.java:150)
        at org.apache.cloudstack.network.lb.CertServiceTest.runUploadSslCertWithCAChain(CertServiceTest.java:143)

runUploadSslCertBadChain(org.apache.cloudstack.network.lb.CertServiceTest)  Time elapsed: 0.011 sec  <<< FAILURE!
java.lang.AssertionError: null
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertTrue(Assert.java:52)
        at org.apache.cloudstack.network.lb.CertServiceTest.runUploadSslCertBadChain(CertServiceTest.java:290)

runUploadSslCertNoRootCert(org.apache.cloudstack.network.lb.CertServiceTest)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: null
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertTrue(Assert.java:52)
        at org.apache.cloudstack.network.lb.CertServiceTest.runUploadSslCertNoRootCert(CertServiceTest.java:342)
```

Any ideas???

**Comment by bhaisaab** [comment]
@swill I'm guessing an environment issue, likely an empty java keystore; what is the version of your maven, java? Are you using Oracle Java or OpenJDK?

This works in my local env that has maven 3.3.3, oracle-java 1.7.0_80 (I had issues with distro provided openjdk 7u95, and I found other folks on ML have reported issues with latest openjdk as well, therefore I had switched to oracle-java).

**Comment by swill** [comment]
The above results are from my CI environment which is running the following:
```
$ mvn --version
Apache Maven 3.0.5 (Red Hat 3.0.5-16)
Maven home: /usr/share/maven
Java version: 1.8.0_77, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-1.8.0-openjdk-1.8.0.77-0.b03.el7_2.x86_64/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.10.0-327.10.1.el7.x86_64", arch: "amd64", family: "unix"
```

I just tried on my laptop which has the following config:
```
$ mvn --version
Apache Maven 3.1.1 (0728685237757ffbf44136acec0402957f723d9a; 2013-09-17 11:22:22-0400)
Maven home: /Users/swill/Dropbox/code/src/apache-maven-3.1.1
Java version: 1.7.0_40, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.7.0_40.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.9.5", arch: "x86_64", family: "mac"
```

When building, I get the following:
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.cloud.network.bigswitch.BigSwitchApiTest
log4j:WARN No appenders could be found for logger (org.apache.commons.httpclient.HttpClient).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Tests run: 36, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.647 sec - in com.cloud.network.bigswitch.BigSwitchApiTest
Running com.cloud.network.bigswitch.BigSwitchBcfUtilsTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.538 sec - in com.cloud.network.bigswitch.BigSwitchBcfUtilsTest
Running com.cloud.network.resource.BigSwitchBcfResourceTest
Tests run: 27, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.084 sec - in com.cloud.network.resource.BigSwitchBcfResourceTest

Results :

Tests run: 64, Failures: 0, Errors: 0, Skipped: 0

[INFO] 
[INFO] --- maven-jar-plugin:2.5:jar (default-jar) @ cloud-plugin-network-bigswitch ---
[INFO] Building jar: /Users/swill/Dropbox/code/cs/cloudstack/plugins/network-elements/bigswitch/target/cloud-plugin-network-bigswitch-4.9.0-SNAPSHOT.jar
[INFO] 
[INFO] --- maven-site-plugin:3.1:attach-descriptor (attach-descriptor) @ cloud-plugin-network-bigswitch ---
[INFO] 
[INFO] >>> maven-pmd-plugin:3.3:check (cloudstack-pmd) @ cloud-plugin-network-bigswitch >>>
[INFO] 
[INFO] --- maven-pmd-plugin:3.3:pmd (pmd) @ cloud-plugin-network-bigswitch ---
[WARNING] Unable to locate Source XRef to link to - DISABLED
[WARNING] Unable to locate Source XRef to link to - DISABLED
Apr 01, 2016 12:41:07 PM net.sourceforge.pmd.PMD removeBrokenRules
WARNING: Removed misconfigured rule: LoosePackageCoupling  cause: No packages or classes specified
Apr 01, 2016 12:41:08 PM net.sourceforge.pmd.PMD removeBrokenRules
WARNING: Removed misconfigured rule: LoosePackageCoupling  cause: No packages or classes specified
Apr 01, 2016 12:41:08 PM net.sourceforge.pmd.PMD removeBrokenRules
WARNING: Removed misconfigured rule: LoosePackageCoupling  cause: No packages or classes specified
Apr 01, 2016 12:41:08 PM net.sourceforge.pmd.PMD removeBrokenRules
WARNING: Removed misconfigured rule: LoosePackageCoupling  cause: No packages or classes specified
Apr 01, 2016 12:41:08 PM net.sourceforge.pmd.PMD removeBrokenRules
WARNING: Removed misconfigured rule: LoosePackageCoupling  cause: No packages or classes specified
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000010ee34f21, pid=38330, tid=12547
#
# JRE version: Java(TM) SE Runtime Environment (7.0_40-b43) (build 1.7.0_40-b43)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (24.0-b56 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# V  [libjvm.dylib+0x234f21]  oopDesc::size()+0x8d
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/swill/Dropbox/code/cs/cloudstack/hs_err_pid38330.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.sun.com/bugreport/crash.jsp
#
Abort trap: 6
```

As suggested, I ran: `ulimit -c unlimited`

After making this change, it builds on my laptop...

**Comment by swill** [comment]
Here is the core dump I had originally when I tried to build on my mac.

[hs_err_pid38330.log.zip](https://github.com/apache/cloudstack/files/200015/hs_err_pid38330.log.zip)

**Comment by bhaisaab** [comment]
@swill based on the output you've shared from your CI env -- `Java version: 1.8.0_77, vendor: Oracle Corporation`; you're using Java 8. Please switch to Java 7 (openjdk or oraclejava) and try again.

**Comment by swill** [comment]
What about this PR forces me to downgrade to Java 7?  My Java 8 build environment has been working fine against master until this PR.

**Comment by swill** [comment]
There were quite a few changes to the way certificates are handled in this PR.  Changes to versions of bouncycastle as well as changing the way the Pem files are handles.  In light of that, the errors I am getting are pretty relevant.

**Comment by jburwell** [comment]
@swill Java8 is not currently supported by the management server.  We know that the process will collapse after running for 24-48 hours.  Part of the issue is that Spring 3 has known issues on Java8.

Per my earlier feedback, I think the Bouncy Castle upgrade should be split into a separate PR.  The issue could be with the changes to the certificate handling code or an issue between the new Bouncy Castle version and one of the transitive dependency updates.  Breaking up this PR would make it much easier to understand whether or not this issue is caused by the dependency upgrades or the code change.

**Comment by swill** [comment]
I have changed the way my CI environment is being built, so it is built to use Java 7 now.  I am testing this PR now, so I will update with status when I have more...

**Comment by swill** [comment]
I get the same behavior with Java 7.

```
$ mvn --version
Apache Maven 3.0.5 (Red Hat 3.0.5-16)
Maven home: /usr/share/maven
Java version: 1.7.0_99, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-1.7.0-openjdk-1.7.0.99-2.6.5.0.el7_2.x86_64/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.10.0-327.10.1.el7.x86_64", arch: "amd64", family: "unix"
```

It fails here:
```
[INFO] Apache CloudStack Framework - Jobs ................ SUCCESS [2.429s]
[INFO] Apache CloudStack Cloud Engine Internal Components API  SUCCESS [1.062s]
[INFO] Apache CloudStack Server .......................... FAILURE [1:01.548s]
[INFO] Apache CloudStack Framework - Quota ............... SKIPPED
[INFO] Apache CloudStack Usage Server .................... SKIPPED
```

When running this:
```
Running org.apache.cloudstack.network.lb.CertServiceTest
Tests run: 13, Failures: 2, Errors: 1, Skipped: 0, Time elapsed: 0.325 sec <<< FAILURE! - in org.apache.cloudstack.network.lb.CertServiceTest
runUploadSslCertWithCAChain(org.apache.cloudstack.network.lb.CertServiceTest)  Time elapsed: 0.05 sec  <<< ERROR!
com.cloud.utils.exception.CloudRuntimeException: Error parsing certificate data Parsing certificate/key failed: Could not parse certificate: java.io.IOException: Empty input
        at org.apache.cloudstack.network.lb.CertServiceImpl.uploadSslCert(CertServiceImpl.java:150)
        at org.apache.cloudstack.network.lb.CertServiceTest.runUploadSslCertWithCAChain(CertServiceTest.java:143)

runUploadSslCertBadChain(org.apache.cloudstack.network.lb.CertServiceTest)  Time elapsed: 0.016 sec  <<< FAILURE!
java.lang.AssertionError: null
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertTrue(Assert.java:52)
        at org.apache.cloudstack.network.lb.CertServiceTest.runUploadSslCertBadChain(CertServiceTest.java:290)

runUploadSslCertNoRootCert(org.apache.cloudstack.network.lb.CertServiceTest)  Time elapsed: 0.023 sec  <<< FAILURE!
java.lang.AssertionError: null
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertTrue(Assert.java:52)
        at org.apache.cloudstack.network.lb.CertServiceTest.runUploadSslCertNoRootCert(CertServiceTest.java:342)
```

We need to better understand why this is happening...

**Comment by bhaisaab** [comment]
@swill can you try again with oracle java 1.7.0_80 there are known
certificate and cipher suite issues causing build issues when using
1.7.0_90+. JDK8 is not supported yet, so you've to use jdk7 for now though
I've tested once and found that ACS mgmt servers works okay with JRE8
(though it is not advised to use it yet).

**Comment by swill** [comment]
@bhaisaab: whether or not we 'can' find a working configuration is not really the point here. I don't think it is acceptable that we break support for `openjdk` and so dramatically reduce the supported environments.  If this goes in as it is, we will break a large percentage of people's environments as well as potentially breaking production environments that use `ooenjdk`.  If we really are going to remove support for `openjdk`, then that needs to be communicated to the community and planned in advance, we can't just make that switch.  We should probably start a thread in dev@ to discuss changing the environment requirements and plan how/when that change happens. This is not something we can just drop on people without proper documentation and an upgrade strategy.

**Comment by bhaisaab** [comment]
@swill I share what worked for me, I'm not saying you have to use oracle-java. The openjdk version 1.7.0_85+ causes test failures for building console proxy in my Linux environment while the same happens on OSX when using 1.7.0_90+, while openjdk version 1.7.0_80 works alright. The issue has to do with a set of cipher suites that got disabled, and random seg fault related issues. Both of them are related to `javac` and not `java`, and they are not consistent across platforms. In production we support openjdk-jre and not the `jdk/javac`, I'm not trying to suggest we remove any sort of java7 jre support at all.

tl;dr -- as I suggested earlier, what you're facing is an environment issue. It it helps, I'll split the PR into three -- one for mvn versions, one for bounty castle and changes by Daan, and one that fixes Travis.

**Comment by bhaisaab** [comment]
@DaanHoogland do you want to have a look, if you can help fix the issue of building against latest openjdk and advise if bountycastle related changes are causing this. Maybe send them as a separate PR?

@swill I've split this PR into two, this one is related to mvn version changes; #1461 will fix the Travis issue

**Comment by DaanHoogland** [comment]
@bhaisaab I'm kind of confused, you write about three parts but I count four. I would agree that the bouncy castle is a completely seperate one. I think mvn changes would include any changes made by me, right? So then we are down to three again.

**Comment by bhaisaab** [comment]
@DaanHoogland feel free to split this however you feel, I've already moved the Travis fix to a separate PR

**Comment by bhaisaab** [comment]
@swill advise how to proceed on this one? are you still getting the build issues
@DaanHoogland ping

**Comment by swill** [comment]
@bhaisaab I am still getting compile errors.  I will reconfigure my CI environment to use an older version of Java to see if that fixes the problem.

```
Tests run: 13, Failures: 2, Errors: 1, Skipped: 0, Time elapsed: 0.477 sec <<< FAILURE! - in org.apache.cloudstack.network.lb.CertServiceTest
runUploadSslCertWithCAChain(org.apache.cloudstack.network.lb.CertServiceTest)  Time elapsed: 0.033 sec  <<< ERROR!
com.cloud.utils.exception.CloudRuntimeException: Error parsing certificate data Parsing certificate/key failed: Could not parse certificate: java.io.IOException: Empty input
        at org.apache.cloudstack.network.lb.CertServiceImpl.uploadSslCert(CertServiceImpl.java:150)
        at org.apache.cloudstack.network.lb.CertServiceTest.runUploadSslCertWithCAChain(CertServiceTest.java:143)

runUploadSslCertBadChain(org.apache.cloudstack.network.lb.CertServiceTest)  Time elapsed: 0.007 sec  <<< FAILURE!
java.lang.AssertionError: null
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertTrue(Assert.java:52)
        at org.apache.cloudstack.network.lb.CertServiceTest.runUploadSslCertBadChain(CertServiceTest.java:290)

runUploadSslCertNoRootCert(org.apache.cloudstack.network.lb.CertServiceTest)  Time elapsed: 0.021 sec  <<< FAILURE!
java.lang.AssertionError: null
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertTrue(Assert.java:52)
        at org.apache.cloudstack.network.lb.CertServiceTest.runUploadSslCertNoRootCert(CertServiceTest.java:342)
```

**Comment by bhaisaab** [comment]
@swill thanks Will, I'll try to split this PR into two to isolate what changes are causing this in your env and see if we can get it fixed/improved separately

**Comment by swill** [comment]
@bhaisaab thank you sir.  sorry this is proving to be a bit complicated.  i appreciate the support.

**Comment by bhaisaab** [comment]
@swill no problem, I'll deal with this during weekends and see how much we can safely get into master.

**Comment by bhaisaab** [comment]
@swill closing this PR, opened two new PRs that split changes from this PR

**Comment by swill** [comment]
Thanks for all the effort @bhaisaab.  :)
